storage/remote/client.go

```diff
@@ -187,3 +190,70 @@ func (c *Client) Read(ctx context.Context, query *prompb.Query) (*prompb.QueryRe
 
 	return resp.Results[0], nil
 }
+
+type labelsResponse struct {
+	Data     []string `json:"data,omitempty"`
+	Error    string   `json:"error,omitempty"`
+	Warnings []string `json:"warnings,omitempty"`
+}
+
+// makeLabelsRequest makes the actual request for /label and /labels/{name}/values from the remote client
+func (c *Client) makeLabelsRequest(ctx context.Context, url string) ([]string, storage.Warnings, error) {
+	httpReq, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "unable to create request")
+	}
+	httpReq.Header.Set("User-Agent", userAgent)
+
+	ctx, cancel := context.WithTimeout(ctx, c.timeout)
+	defer cancel()
+
+	httpResp, err := c.client.Do(httpReq.WithContext(ctx))
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "error sending request")
+	}
+	defer func() {
+		io.Copy(ioutil.Discard, httpResp.Body)
+		httpResp.Body.Close()
+	}()
+
+	body, err := ioutil.ReadAll(httpResp.Body)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, fmt.Sprintf("error reading response. HTTP status code: %s", httpResp.Status))
+	}
+
+	if httpResp.StatusCode/100 != 2 {
+		return nil, nil, errors.Errorf("remote server %s returned HTTP status %s: %s", url, httpResp.Status, strings.TrimSpace(string(body)))
+	}
+
+	var resp labelsResponse
+	err = json.Unmarshal(body, &resp)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "unable to unmarshal response body")
+	}
+
+	if resp.Error != "" {
+		return nil, nil, errors.New(resp.Error)
+	}
+
+	var warnings storage.Warnings
+	for _, warning := range resp.Warnings {
+		warnings = append(warnings, errors.New(warning))
+	}
+
+	return resp.Data, warnings, nil
+}
+
+// LabelValues queries label values from a remote endpoint.
+func (c *Client) LabelValues(ctx context.Context, name string) ([]string, storage.Warnings, error) {
+	re := regexp.MustCompile(`read/?$`)
```

Comment:
This looks like a code smell. Can we create client with correct URLs once to not rely on computing regexp dynamically every time? Also, can we use https://github.com/prometheus/client_golang/blob/master/api/prometheus/v1/api.go#L235